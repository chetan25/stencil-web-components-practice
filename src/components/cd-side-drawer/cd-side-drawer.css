:host {
  display: block;
}
aside {
  position: fixed;
  width: var(--drawer-width , 30rem);
  max-width: 80%;
  height: 100vh;
  background: var(--drawer-background, #faf7f7);
  box-shadow: 0 2px 8px rgba(0,0,0,0.26);
  transition: all 0.3s ease-out;
  z-index: var(--drawer-zIndex, 20);
}

:host([slide='left']) aside {
  top: 0;
  left: -100%;
}

:host([slide='right']) aside {
  top: 0;
  right: -100%;
}

:host([slide='left'][opened]) aside {
  left: 0;
}

:host([slide='right'][opened]) aside {
  right: 0;
}

.backdrop {
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100vh;
  background: rgba(0,0,0,0.75);
  opacity: 0;
  z-index: 10;
  pointer-events: none;
  transition: opacity 0.3s ease-out;
}

:host([opened]) .backdrop {
  opacity: 1;
  pointer-events: all;
}

header {
  padding: 1rem;
  background: var(--drawer-header, black);
  position: relative;
}
header slot {
  font-size: var(--drawer-header-font-size , 1.5rem);
  color: var(--drawer-header-color ,white);
  margin: 0;
}

header .close {
  position: absolute;
  top: 0;
  right: 0;
  padding: 1rem;
  color: white;
  background: transparent;
  font-size: 1.5rem;
  border: none;
}

header button:focus,
header button:hover {
  outline: none;
  cursor: pointer;
}

::slotted(*) {
  /*color: purple;*/
  font-family: sans-serif;
  font-size: medium;
  font-weight: bold;
}

::slotted([slot="menu"]) {
  /* Apply styles to non-empty menu */
}
