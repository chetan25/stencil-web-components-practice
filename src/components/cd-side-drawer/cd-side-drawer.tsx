import { Component, Prop, Method, Element,  h } from '@stencil/core';

export type Slide = 'left' | 'right';

@Component({
  tag: 'cd-side-drawer',
  styleUrl: './cd-side-drawer.css',
  shadow: true
  // scoped: true, //tells stencil that all css are scoped to this file(emulation of shadow dom)
})
export class CdSideDrawer {
  /**
   * Only used in css to change styling
   * by default props are immutable inside component,
   * to make them mutable use mutable
   */
  @Prop({reflect: true, mutable: true}) opened: boolean = false;

  /**
   * Prop use that governs the direction of the Drawer slide
   */
  @Prop({reflect: true, mutable: true}) slide: Slide = 'left';

  /**
   * Method to open Drawer
   */
  @Method()
  onOpen() {
    this.opened = true;
  }

  /**
   * Method to close Drawer
   */
  @Method()
  onClose() {
    this.opened = false;
  }

  @Element() hostElement: HTMLElement;

  onCloseDrawer = () => {
    this.opened = false;
  }

  // <slot/> could be style from outside
  render() {
    return [
      <div class="backdrop" onClick={this.onCloseDrawer}></div>,
      <aside>
        <header>
          <slot name="drawer-header" />
          { !!this.hostElement.querySelector('[slot="drawer-close-icon"]')
             ? <slot name="drawer-close-icon" />
             : <button class='close' onClick={this.onCloseDrawer}>X</button>
          }
        </header>
        <main>
          <slot name="drawer-content" />
        </main>
      </aside>
    ];
  }
}
