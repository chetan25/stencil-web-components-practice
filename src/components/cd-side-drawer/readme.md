# cd-side-drawer



<!-- Auto Generated Below -->


## Properties

| Property | Attribute | Description                                                                                                          | Type                | Default     |
| -------- | --------- | -------------------------------------------------------------------------------------------------------------------- | ------------------- | ----------- |
| `opened` | `opened`  | Only used in css to change styling by default props are immutable inside component, to make them mutable use mutable | `boolean`           | `undefined` |
| `slide`  | `slide`   | Prop use that governs the direction of the Drawer slide                                                              | `"left" \| "right"` | `'left'`    |


## Methods

### `onClose() => Promise<void>`

Method to close Drawer

#### Returns

Type: `Promise<void>`



### `onOpen() => Promise<void>`

Method to open Drawer

#### Returns

Type: `Promise<void>`




----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
