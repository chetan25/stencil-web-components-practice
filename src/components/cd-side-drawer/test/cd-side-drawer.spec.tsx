import { newSpecPage } from '@stencil/core/testing';
import { CdSideDrawer } from '../cd-side-drawer';

describe('cd-side-drawer', () => {
  it('renders CdSideDrawer with shadow DOM', async () => {
    const html = `
     <cd-side-drawer slide="right">
       <div slot="drawer-header" class="header-class">Settings</div>
     <div slot="drawer-content"></div>
    </cd-side-drawer>
  `;
    const page = await newSpecPage({
      components: [CdSideDrawer],
      html: html,
    });

    expect(page.root.shadowRoot).toBeTruthy();
    expect(page.root.querySelector('.backdrop')).toBeFalsy();
    expect(page.root.shadowRoot.querySelector('.backdrop')).toBeTruthy();
    expect(page.root.querySelector('.header-class')).toBeTruthy();
    expect(page.root.shadowRoot.querySelector('.header-class')).toBeFalsy();

    expect(page.rootInstance.slide).toEqual('right');
    // expect(page.root).toMatchSnapshot();
  });

  it('renders all passed slots element', async () => {
    const html = `
     <cd-side-drawer>
       <div slot="drawer-header" class="header-class">Settings</div>
       <button
          slot="drawer-close-icon"
          class='close'
          id="sidebarClose"
       >Close</button>
       <div slot="drawer-content" id="content"><h2>Content</h2></div>
    </cd-side-drawer>
  `;
    const page = await newSpecPage({
      components: [CdSideDrawer],
      html: html,
    });

    expect(page.root.querySelector('.header-class').textContent).toEqual('Settings');
    expect(page.root.querySelector('.close').textContent).toEqual('Close');
    expect(page.root.querySelector('#content').textContent).toEqual('Content');
    expect(page.root.shadowRoot.querySelector('.close')).toBeFalsy();

    expect(page.rootInstance.slide).toEqual('left');
    expect(page.rootInstance.opened).toEqual(false);
  });

  it('renders default slots element', async () => {
    const html = `
     <cd-side-drawer slide="left">
       <div slot="drawer-header" class="header-class">Settings</div>
       <div slot="drawer-content" id="content"><h2>Content</h2></div>
    </cd-side-drawer>
  `;
    const page = await newSpecPage({
      components: [CdSideDrawer],
      html: html,
    });

    expect(page.root.querySelector('.header-class').textContent).toEqual('Settings');
    expect(page.root.querySelector('.close')).toBeFalsy();
    expect(page.root.shadowRoot.querySelector('.close').textContent).toEqual('X');
  });

  it('onCLose and onOpen methods are fired properly', async () => {
    const html = `
     <cd-side-drawer slide="left">
       <div slot="drawer-header" class="header-class">Settings</div>
       <div slot="drawer-content" id="content"><h2>Content</h2></div>
    </cd-side-drawer>
  `;
    const page = await newSpecPage({
      components: [CdSideDrawer],
      html: html,
    });
    await page.rootInstance.onOpen();
    expect(page.rootInstance.opened).toEqual(true);

    await page.rootInstance.onClose();
    expect(page.rootInstance.opened).toEqual(false);
  });
});
