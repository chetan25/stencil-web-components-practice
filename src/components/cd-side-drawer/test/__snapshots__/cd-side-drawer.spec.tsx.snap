// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cd-side-drawer renders 1`] = `
<cd-side-drawer slide="left">
  <mock:shadow-root>
    <div class="backdrop"></div>
    <aside>
      <header>
        <slot name="drawer-header"></slot>
        <button class="close">
          X
        </button>
      </header>
      <main>
        <slot name="drawer-content"></slot>
      </main>
    </aside>
  </mock:shadow-root>
  <div class="header-class" slot="drawer-header">
    Settings
  </div>
  <div slot="drawer-content">
    <div>
      Content
    </div>
  </div>
</cd-side-drawer>
`;
